streams/src/main/java/org/apache/kafka/streams/state/OffsetCheckpoint.java

```diff
@@ -55,7 +56,6 @@
     private final Object lock;
 
     public OffsetCheckpoint(File file) throws IOException {
-        new File(file + ".tmp").delete(); // try to delete any existing temp files for cleanliness
```

Comment:
Let me elaborate on this a bit more.

It is unsafe to create multiple instances of OffsetCheckpoint for the same path and invoke `write` concurrently since each instance has its own lock. We could improve this slightly by ensuring that the same lock is used for a given path within a single JVM, but it doesn't help with the case where the tests are run over multiple JVMs (which we do by default). For that, we need to ensure that the filenames never conflict in the first place (as you said).

Thankfully, we already do that within our tests. The tests use `TestUtils.tempDir()` for `log.dir`, which in turn, calls `Files.createTempDirectory`. The spec for `createTempDirectory` says it returns "the path to the newly created directory that did not exist before this method was invoked". The implementation uses a `SecureRandom` to generate the path and tries to create the directory in a loop until `FileAlreadyExistsException` is not thrown. The current implementation of `TestUtils.tempDir` was introduced in KAFKA-2718 because the previous one less robust. Having said that, KAFKA-2718 was reopened because tests still seem to fail occasionally with similar errors as before it was fixed (although it seems to happen more rarely) and more investigation is needed.

KAFKA-3063 was caused by two instances of `OffsetCheckpoint` in the same test (one within the broker and one within the test), but the test instance was only used for reading (as said earlier, writing concurrently is still unsafe now). As one data point (with the usual caveats that go with it), I ran several Jenkins builds (over 10) and I did not see any case where the temporary directory used by test that caused the JVM to be killed was used by any other test.